lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+var errServicesStopFailed = errors.New("services stop failed")
+
+type ephemeralService interface {
+	Run() error
+	Stop() error
+}
+
+type finalizationHandler struct {
+	servicesLock       sync.Mutex
+	finalizationEngine ephemeralService
+	votingRound        ephemeralService
+
+	newServices   func() (engine, voting ephemeralService)
+	initiateRound func() error
+
+	stopCh      chan struct{}
+	handlerDone chan struct{}
+	errorCh     chan error
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	// builder is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	builder := func() (engine, voting ephemeralService) {
+		finalizationEngine := newFinalizationEngine(service)
+		votingRound := newHandleVotingRound(service, finalizationEngine.actionCh)
+		return finalizationEngine, votingRound
+	}
```

Comment:
nit it would be cleaner imo to just have `*Service` as a field in the handler, and inline the content of that `builder` within `waitServices`

Reply:
I was suggesting to create the ephemeral services inline within `runEphemeralServices`, and without a functional field. However, I guess you use that functional field to mock it for tests right?